### DISCUSSION TITLE
Gradle Incremental build with TestContainers + Liquibase + JOOQ Codegen

### DISCUSSION THREAD

**Discussion by jobayle** [jOOQ/jOOQ#17744] [Q&A] [answered]
Dears, I tried to enable caching on my CodegenTask tasks, but can't get it to work...

The goal is to avoid starting support containers to re-run liquibase update and jooq codegen and on every run (a waste of time and resources).

I'm following: https://www.jooq.org/doc/latest/manual/code-generation/codegen-gradle/codegen-gradle-data-change-management/

But as I have several executions I also tried the following:

```kotlin
tasks.withType<CodegenTask>().configureEach {
    inputs.files(fileTree("src/main/resources/liquibase"))
}
```

Here's a minimal reproducer that copies my use case (multiple databases, multiple executions of jooq codegen):
https://github.com/jobayle/jooq-incremental
(there's a failing test on purpose)

Every rerun of `./gradlew build` containers are started, liquibase creates the schema and jooq regenerate model object from the schema.

Also just running task `test` after fixing the failing test reruns everything (containers, liquibase, codegen).

What is wrong in my build script?
Can you get it to work?

Thanks!

**Comment by lukaseder** [comment, marked as answer]
If you run your  build with `--info` logging, you can see:

```
> Task :jooqCodegenDb01
Caching disabled for task ':jooqCodegenDb01' because:
  Build cache is disabled
Task ':jooqCodegenDb01' is not up-to-date because:
  Value of input property 'configuration' has changed for task ':jooqCodegenDb01'
The input changes require a full rebuild for incremental task ':jooqCodegenDb01'.
```

This is because of your section:

```kotlin
// Configure JOOQ
project.jooq.executions.named("db01").get().configuration.jdbc.url = url + "db01"
project.jooq.executions.named("db02").get().configuration.jdbc.url = url + "db02"
```

You mustn't dynamically change the jOOQ configuration for caching to work. There are other ways to pass this information to jOOQ, including using system properties:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-system-properties/

You'll have to do the same with liquibase and the container startup.

I guess it would help to allow for passing the property name for later lookup, rather than just allowing a single, hard-coded property name:

- https://github.com/jOOQ/jOOQ/issues/17752

**Comment by jobayle** [comment]
Using system properties, it works, even though the code has to work around the limitation of a single property for multiple invocations:
https://github.com/jobayle/jooq-incremental/commit/f79bd468d10a0f3a3c72f6151ae489d0b99f438e

Looking forward to replacing that with `urlProperty` as soon as it is released.

Thank you!
